informer/informer.go

```diff
@@ -0,0 +1,383 @@
+// That the resync period configured for the informer will trigger periodic
+// updates of event objects via the update channel.
+func (i *Informer) Watch(ctx context.Context, endpoint string, factory ZeroObjectFactory) (chan watch.Event, chan watch.Event, chan error) {
+	done := make(chan struct{}, 1)
+	eventChan := make(chan watch.Event, 1)
+
+	deleteChan := make(chan watch.Event, 1)
+	updateChan := make(chan watch.Event, 1)
+	errChan := make(chan error, 1)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+			case event := <-eventChan:
+				switch event.Type {
+				case watch.Added:
+					err := i.cacheAndSendIfNotExists(event, updateChan)
+					if err != nil {
+						errChan <- microerror.Mask(err)
+					}
+				case watch.Deleted:
+					err := i.uncacheAndSend(event, deleteChan)
+					if err != nil {
+						errChan <- microerror.Mask(err)
+					}
+				case watch.Modified:
+					err := i.cacheAndSend(event, updateChan)
+					if err != nil {
+						errChan <- microerror.Mask(err)
+					}
+				default:
+					errChan <- microerror.Maskf(invalidEventError, "%#v", event)
+				}
+			}
+		}
+	}()
+
+	go func() {
+		// Here we fill the informer cache initially and release the event objects
+		// the very first time after the program started. This is a special case and
+		// guarantees any configured rate limitting is properly done.
+		{
+			err := i.fillCache(ctx, endpoint, factory, eventChan)
+			if err != nil {
+				errChan <- microerror.Mask(err)
+			}
+			close(i.initializer)
+			i.sendCachedEvents(ctx, updateChan)
+		}
+
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				ctx, cancelFunc := context.WithCancel(ctx)
+				go func() {
+					for {
+						select {
+						case <-ctx.Done():
+							return
+						default:
+							err := i.streamEvents(ctx, endpoint, factory, eventChan)
+							if err != nil {
+								errChan <- microerror.Mask(err)
+							}
+						}
+					}
+				}()
+
+				<-time.After(i.resyncPeriod)
+
+				i.sendCachedEvents(ctx, updateChan)
+				cancelFunc()
+			}
+		}
+	}()
+
+	go func() {
+		<-ctx.Done()
+
+		close(done)
+		close(eventChan)
+
+		close(deleteChan)
+		close(updateChan)
+		close(errChan)
+	}()
+
+	return deleteChan, updateChan, errChan
+}
+
+// cacheAndSend sends the provided event object to the provided update channel
+// and caches the provided event object.
+func (i *Informer) cacheAndSend(event watch.Event, updateChan chan watch.Event) error {
+	updateChan <- event
+
+	k, err := cache.MetaNamespaceKeyFunc(event.Object)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	i.cache.Store(k, event)
+
+	return nil
+}
+
+// cacheAndSendIfNotExists handles watch.ADDED events. These events can happen
+// because of different reasons.
+//
+//     - The watcher may receives a new event object because a new object was
+//       created in the API.
+//     - The watcher may syncs the very first time on programm start.
+//     - The watcher may resyncs and receives an event object we already know.
```

Comment:
There is only an update channel which receives create/update events. Objects are replayed after resync period and our create/update mechanism is being executed over and over again. All good from my point of view. Maybe check the integration tests. This PR is already outdated because there happened quite a lot of changes meanwhile. Thanks for your feedback though.